enhancements/rhcos/split-rhcos-into-layers.md

```diff
@@ -0,0 +1,458 @@
+but it would be wasteful to not symlink the bootimages on the server side to
+share across OCP versions that share the same set of bootimages.
+
+As such, the versioning change should be transparent to end-users, but on the
+off chance that a customer is relying on the current semantics, we should still
+announce it in the GA release notes.
+
+#### Hidden dependencies
+
+Given that we're effectively removing packages from the bootimages, we may
+uncover hidden dependencies on the removed packages both within other components
+of OpenShift and in end-user code.
+
+Regarding other OCP components, the goal is to catch these dependencies during
+CI testing of the installer PR that switches over to the new bootimages. We
+need to raise awareness of the change to the whole OCP org and call for extra
+testing. This needs to be a coordinated event with ART, DPTP, and TRT. Of
+course, even after merging, if any regression is found, we can easily revert the
+PR to get back to green.
+
+Regarding end-user code, although bootimages are by their very nature
+transient (we boot, and shortly after pivot into the node image), there are
+some situations where user-provided code may run (e.g. when manually extending
+`bootstrap.ign` or `master.ign` or when scripting the live ISO). This code may
+be relying on packages that are going away (e.g. some local-only `oc` command,
+or grepping `--help` texts of the kubelet, etc...).
+
+We shouldn't worry *too much* about breaking these use cases, but it still seems
+worth highlighting in the GA release notes that these OCP packages are no longer
+present.
+
+### Drawbacks
+
+- This will make it somewhat harder (or at the very least much more explicit) to
+  fast-track RHEL fixes into specific OCP releases. Because everything beneath
+  the node layer is RHEL-versioned and possibly shared across multiple OCP
+  releases, any fast-tracks would need to happen by doing an RPM upgrade within
+  the node layer. This is more cumbersome than just tagging packages into the
+  the ART plashets as currently done, but makes us more thorough in our known
+  deltas from RHEL.
+  - We should do this as a last resort and instead work more closely with RHEL
+    package maintainers to get fixes into the releases we need.
+
+## Open Questions [optional]
+
+- How to handle the early kernels we currently get from RHEL?
+  - The early kernels are RHEL-versioned and it would make sense to have them be
+    part of the CoreOS layer. (IOW, we always use the same early kernels for the set
+    of OCP streams sharing the same RHEL version.)
+  - Currently, this is done by tagging the kernel builds into a specific tag in
+    Brew and those builds then being added to the ART plashets. To avoid having
+    to source the (OCP-versioned) ART plashets during the CoreOS builds, we
+    could instead create new repos for these packages that are clearly defined
+    to be RHEL-versioned. The CoreOS build could then add that repo alongside
```

Comment:
I think ART would need to dnf upgrade the kernel in the node layer from their plashet. If the coreos layer is up-to-date, it is a no-op. Otherwise, we lose some of the benefit of the layer decoupling.

Reply:
Hmm, when would it be not up to date? The ART plashet sources it from the same place the RHCOS image would source it from. It's certainly possible to replace the kernel in the node image build, but IMO it's much more appropriate to respin the RHCOS base image for it.

Reply:
> The ART plashet sources it from the same place the RHCOS image would source it from

Seems like I misinterpreted "we could instead create new repos for these packages that are clearly defined to be RHEL-versioned" which sounded like you would not be consuming ART generated repos. Reading it now, I think you mean that the source would still be created by ART, but RHEL-version specific (e.g. plashets/el9.2 instead of plashets/4.19). 

This would mean we have no ability to have different kernel versions between different Z's using the same RHEL (tagging in a kernel fix for 4.x could impact RHCOS for 4.y). @joepvd concerns here?

In any case, ART needs to ability to create hotfixes including kernel changes. This would need to be layered on top of the RHCOS image unless RHCOS will allow us to trigger custom builds as they do today.

Reply:
> Reading it now, I think you mean that the source would still be created by ART, but RHEL-version specific (e.g. plashets/el9.2 instead of plashets/4.19).

Yes, exactly. I'll clarify this.

> This would mean we have no ability to have different kernel versions between different Z's using the same RHEL (tagging in a kernel fix for 4.x could impact RHCOS for 4.y).

To me, that's a feature, not a bug. :) Differing from RHEL should be the exception not the rule, so having that tracked in the layered image instead only when needed seems appropriate.

> In any case, ART needs to ability to create hotfixes including kernel changes. This would need to be layered on top of the RHCOS image unless RHCOS will allow us to trigger custom builds as they do today.

Ahh yes, thanks for bringing up hotfixes. I think in general the pipecfg hotfix path can be considered deprecated by layering. So instead, you'd do a custom layered build where you can e.g. apply hotfixes for whatever packages you want there. The exception is if it's a hotfix required for firstboot though I suspect that's much less common. (But even then, long-term us or users can transform the layered image into bootimages.) I'll mention this.